MAISTRA-2590 Add readiness probe to operator

luksa opened the pull request.

luksa committed:
MAISTRA-2590 Add readiness probe to operator

rcernich commented:
Will this return ready regardless of leader status?  I think there was an issue very early on, where a Deployment update failed rollout, as the new pod would never become ready until after the old pod was deleted, which wouldn't happen until the new pod was ready.  This was long ago, and my memory could be fuzzy.

luksa commented:
I'll have to check. Typically, a controller(manager) should not even run until it becomes the leader (and should terminate immediately when it loses the leader status). That said, I do remember seeing istiod doing stuff even before it became the leader. I'll check.

luksa commented:
When I modify the deployment, the old pod is replaced with the new one with no issues, so it looks like this works as it should. I'll  do some further digging to see if the controller is doing anything that it shouldn't be (while not the leader).

luksa committed:
Prevent deadlock during operator update

When a rolling update of the operator is performed, the new pod never becomes ready because it doesn't start the webhook server until it becomes the leader. It can't become the leader until the old pod terminates. But in a rolling update, the old pod isn't terminated until the new pod becomes ready.

luksa commented:
Ignore my last comment. I was testing istiod instead of the operator (note to self: don't code or test during a meeting). 

It turns out that adding the readiness probe does in fact break rolling updates of the operator. The webhook server in the operator, which the readiness probe uses, does not start until the operator instance becomes the leader. During a rolling update, the new instance must become ready before the old instance is deleted. Because the old instance never releases the leader lease, the new instance never becomes the leader and thus never becomes ready. 

This is tricky. We could modify the operator to run the webhook server even when it isn't the leader. The main reason for using leader election is to prevent multiple operators from reconciling the same objects. Two operators handling webhook requests is not really problematic, as each webhook request would only be processed by one instance. However, in order for the webhook server to run, the operator needs to run its local object cache, which means it needs to open a bunch of watch connections to the API server. Also, during an operator upgrade, two different versions of the operator would be handling validation/mutation requests. This could potentially result in some weird, although transient issues. 

I've solved the deadlock issue by changing the operator's update strategy from `RollingUpdate` to `Recreate`. This way, the old operator instance is shut down, releasing the leader lock, before the new instance is started. This new instance can become the leader and thus become ready. The downside is a short time window during which no CRUD operations on SMCP/SMMR/SMM resources can be performed and no reconciliation takes place. I don't see this as a problem, because operator updates aren't frequent and one can easily refrain from modifying SMCP/SMMR/SMM resources during that time. 

The alternative solution, where some parts of the operator would be active even when the operator isn't the leader, seems much more dangerous.

dgn commented:
Thanks Marko, I think that approach makes a lot of sense. I agree that activating some parts of the operator without being the leader is dangerous. And as the operator is mostly reconciling persistent resources, a Recreate strategy will be fine- of course, you'll have the issue that during upgrades, you can have transient errors creating maistra resources because the webhook might not be available, but the same problem exists today (just for a slighly shorter duration, because the new leader is already started when the old is shutdown, and in this scenario, we'll still have to wait for ContainerCreate to complete).

My only worry here is that we're reporting as ready before the webhook CA certs have been updated- so we might still run into issues during e.g. automated tests, right? If we just wait for the operator to be ready and immediately create resources, we might see 403s - but maybe that time span is so short that it doesn't matter?

dgn commented:
Thinking about it, would it make sense maybe to introduce another state, e.g. `reconciling`? That way, we could use Ready just to signal that we're up and running, and keep using rolling upgrades. And when we're the leader, serving all endpoints and have updated all webhooks, we can report `reconciling` which other components could wait on

jwendell commented:
"Ready" is the standard, it's what users recognize. IMO we should stick
with it.

On Tue, Sep 28, 2021 at 10:26 AM Daniel Grimm ***@***.***>
wrote:

> Thinking about it, would it make sense maybe to introduce another state,
> e.g. reconciling? That way, we could use Ready just to signal that we're
> up and running, and keep using rolling upgrades. And when we're the leader,
> serving all endpoints and have updated all webhooks, we can report
> reconciling which other components could wait on
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/maistra/istio-operator/pull/860#issuecomment-929282947>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA6WP3W7553MPD6M2K4MKDUEHGAJANCNFSM5ERQMKSA>
> .
>


-- 
Jonh Wendell
https://bani.com.br

ocofaigh commented:
Hi folks, any conclusion on what way to handle this? It seems even with a 10min sleep after deploying the Service Mesh operator and before applying the ServiceMeshControlPlane to try and workaround the missing readiness probe, we are still reproducing:

```
Error: Internal error occurred: failed calling webhook "smcp.mutation.maistra.io": Post "https://maistra-admission-controller.openshift-operators.svc:443/mutate-smcp?timeout=10s": dial tcp 172.17.61.133:11999: connect: connection refused
```

luksa commented:
@ocofaigh Are you using the same approach as in this PR? You added a readiness probe _and_ changed the Deployment's update strategy to `Recreate`?

ocofaigh commented:
@luksa No, I simply added a sleep in our scripts after deploying the Service Mesh operator and before applying the ServiceMeshControlPlane.

luksa commented:
@ocofaigh The first time you run the operator, the container restarts after 30s because some secrets are missing (but the operator creates these secrets during the first startup). When it then restarts, the secrets exist and the operator starts up fully in a few seconds. If this doesn't happen in your case, check the operator's log. It probably can't become the leader for some reason.

jwendell commented:
@luksa This needs a rebase.

Do we still want this in for 2.1.1? Or this is a thing for 2.2?

luksa commented:
Best to defer to 2.2 for now.

dgn commented:
FYI: the default branch has been updated to `maistra-2.2`. please make sure that this change lands on both `maistra-2.1` and `maistra-2.2`. you can use the `/cherry-pick <branch>` command

luksa commented:
/hold #888 must be merged first.

luksa committed:
Make gen

rcernich commented on pkg/controller/servicemesh/webhooks/server.go:
```diff
@@ -82,5 +86,13 @@ func Add(mgr manager.Manager) error {
 	})
 
+	log.Info("Adding readiness probe handler")
+	readinessProbeHandler := func(response http.ResponseWriter, request *http.Request) {
```
Does this end up waiting until the webhook CAs are updated, or is that no longer necessary?

dgn commented:
/cherry-pick maistra-2.3

dgn commented:
Can we revive this PR? I think it would be really useful- but I think we should address @rcernich's comment, ie make sure we don't report readiness until CA bundles have been updated

ocofaigh commented:
Any plans to revisie this PR?

luksa commented:
@ocofaigh yes, I hope to find time for this next week so it's released in OSSM 2.4.0.

luksa commented:
Closing. Replaced by #1151
